examples/model_compress/pruning/taylorfo_pruning_with_ddp_torch.py

```diff
@@ -0,0 +1,257 @@
+    criterion: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    lr_scheduler: torch.optim.lr_scheduler._LRScheduler = None,
+    max_steps: int = None, max_epochs: int = None,
+    local_rank: int = -1,
+    save_best_model: bool = False, save_path: str = None,
+    log_path: str = None,
+    evaluation_func=None,
+):
+    model.train()
+    current_step = 0
+    best_acc = 0.
+
+    for current_epoch in range(max_epochs if max_epochs else 2):
+        for (data, target) in training_dataloader:
+            data, target = data.cuda(), target.cuda()
+            optimizer.zero_grad()
+            output = model(data)
+            loss = criterion(output, target)
+
+            loss.backward()
+            optimizer.step()
+
+            if lr_scheduler:
+                lr_scheduler.step()
+
+            current_step += 1
+
+        # evaluation for every 1000 steps
+        if current_step % 1000 == 0 or current_step % len(training_dataloader) == 0:
+            acc = evaluation_func(model)
+            with (log_path).open('a+') as f:
+                msg = '[{}] Epoch {}, Step {}: Acc: {} Loss:{}\n'.format(time.asctime(time.localtime(time.time())), \
+                    current_epoch, current_step, acc, loss.item())
+                f.write(msg)
+            if save_best_model and best_acc < acc:
+                assert save_path is not None
+                if local_rank == 0:
+                    torch.save(model.module.state_dict(), save_path)
+                best_acc = acc
+
+        if max_steps and current_step >= max_steps:
+            return best_acc
+
+    return best_acc
+
+def evaluation(validation_dataloader: DataLoader, model: nn.Module):
+    training = model.training
+    model.eval()
+
+    correct = 0.0
+    with torch.no_grad():
+        for data, target in validation_dataloader:
+            data, target = data.cuda(), target.cuda()
+            output = model(data)
+            pred = output.argmax(dim=1, keepdim=True)
+            correct += pred.eq(target.view_as(pred)).sum().item()
+    acc = 100 * correct / len(validation_dataloader.dataset)
+    # average acc in different local_ranks
+    average_acc = torch.tensor([acc]).cuda()
+    dist.all_reduce(average_acc, op=dist.ReduceOp.SUM)
+    world_size = dist.get_world_size()
+    average_acc = average_acc / world_size
+
+    print('Average Accuracy: {}%\n'.format(average_acc.item()))
+    model.train(training)
+
+    return average_acc.item()
+
+def optimizer_scheduler_generator(model, _lr=0.1, _momentum=0.9, _weight_decay=5e-4, total_epoch=160):
+    optimizer = torch.optim.SGD(model.parameters(), lr=_lr, momentum=_momentum, weight_decay=_weight_decay)
+    scheduler = MultiStepLR(optimizer, milestones=[int(total_epoch * 0.5), int(total_epoch * 0.75)], gamma=0.1)
+    return optimizer, scheduler
+
+def retrain_model(
+    args,
+    model: nn.Module = None,
+):
+    # create an ddp model
+    if model is None:  # pretraining process
```

Comment:
why pretrain is in retrain?

Reply:
the retrain() can be used in the pre-training process and fine-tuning process, so it is uniformly named retrain.